Improve corner cases of deleteat!

bkamins opened the pull request:
This addresses issues discussed in https://github.com/JuliaLang/julia/issues/42065#issuecomment-910729447 for `deleteat!`.

bkamins committed:
Improve corner cases of deleteat!

bkamins committed:
Update array.jl

bkamins commented on base/array.jl:
```diff
@@ -1517,6 +1527,7 @@ function _deleteat!(a::Vector, inds, dltd=Nowhere())
         y = iterate(inds, s)
         y === nothing && break
         (i,s) = y
+        i isa Bool && depawrn("passing Bool as an index is deprecated", deleteat!)
```
```suggestion
```

bkamins committed:
Update base/array.jl

bkamins committed:
Update bitarray.jl

bkamins commented:
What problems this PR tries to solve:
```
julia> deleteat!([1, 2, 3], true)
2-element Vector{Int64}:
 2
 3

julia> deleteat!([1, 2, 3], true:true)
2-element Vector{Int64}:
 2
 3

julia> deleteat!(trues(3), [true, false, false])
ERROR: ArgumentError: indices must be unique and sorted

julia> deleteat!(trues(3), [UInt8(1)])
ERROR: MethodError: no method matching copy_chunks!(::Vector{UInt64}, ::UInt8, ::Vector{UInt64}, ::Int64, ::Int64)
```

bkamins commented on base/array.jl:
```diff
@@ -1462,12 +1462,22 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 2)
-    isempty(r) || _deleteat!(a, first(r), length(r))
-    return a
+    if eltype(r) isa Bool
+        return invoke(deleteat!, Tuple{Vector, AbstractVector{Bool}, a, r)
```
```suggestion
        return invoke(deleteat!, Tuple{Vector, AbstractVector{Bool}}, a, r)
```

bkamins committed:
Update base/array.jl

bkamins commented:
I will add tests and NEWS.md entry if the proposed changes would be acceptable.

bkamins commented on base/bitarray.jl:
```diff
@@ -1019,6 +1021,13 @@ function deleteat!(B::BitVector, inds)
+    if length(B) != length(inds)
+        throw(ArgumentError("length of Bool indices inds must match the length of B"))
+    end
+    return deleteat!(B, findall(inds))
```
clearly this can be made faster, but will lead to some code duplication. Should I improve things here?

JeffBezanson replied:
Seems fine for now. Would it be faster to reuse the code for Vector (at least avoiding the allocation)? The Vector method should work if `_deleteend!` is just replaced with `resize!`.

bkamins replied:
no - it would not be fully efficient. I will implement an efficient method and you will judge.

dkarrasch commented on base/array.jl:
```diff
@@ -1462,12 +1462,22 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 2)
 """
-deleteat!(a::Vector, i::Integer) = (_deleteat!(a, i, 1); a)
+function deleteat!(a::Vector, i::Integer)
+    i isa Bool && depawrn("passing Bool as an index is deprecated", deleteat!)
```
```suggestion
    i isa Bool && depwarn("passing Bool as an index is deprecated", deleteat!)
```

dkarrasch commented on base/array.jl:
```diff
@@ -1462,12 +1462,22 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 2)
+    else
+        n = length(a)
+        f = first(r)
+        f isa Bool && depawrn("passing Bool as an index is deprecated", deleteat!)
```
```suggestion
        f isa Bool && depwarn("passing Bool as an index is deprecated", deleteat!)
```

dkarrasch commented on base/bitarray.jl:
```diff
@@ -947,6 +947,7 @@ function _deleteat!(B::BitVector, i::Int)
 end
 
 function deleteat!(B::BitVector, i::Integer)
+    i isa Bool && depawrn("passing Bool as an index is deprecated", deleteat!)
```
```suggestion
    i isa Bool && depwarn("passing Bool as an index is deprecated", deleteat!)
```

dkarrasch commented on base/bitarray.jl:
```diff
@@ -992,13 +993,14 @@ function deleteat!(B::BitVector, inds)
     y = iterate(inds, s)
     while y !== nothing
         (i, s) = y
+        i isa Bool && depawrn("passing Bool as an index is deprecated", deleteat!)
```
```suggestion
        i isa Bool && depwarn("passing Bool as an index is deprecated", deleteat!)
```

dkarrasch reviewed (commented):
Just stumbled across the typo.

bkamins committed:
Apply suggestions from code review

Co-authored-by: Daniel Karrasch <daniel.karrasch@posteo.de>

JeffBezanson commented on base/array.jl:
```diff
@@ -1462,12 +1462,22 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 2)
-    n = length(a)
-    isempty(r) || _deleteat!(a, first(r), length(r))
-    return a
+    if eltype(r) isa Bool
```
```suggestion
    if eltype(r) === Bool
```

bkamins committed:
Update base/array.jl

Co-authored-by: Jeff Bezanson <jeff.bezanson@gmail.com>

bkamins committed:
implement faster deleteat! and initial tests

bkamins commented:
I have not made PRs to Julia for some time. Could someone please guide me through which CI jobs are running the actual tests of the functionality? (as there seem to be many jobs doing some specific testing) Thank you!

bkamins committed:
add tests

bkamins committed:
add tests

nalimilan commented:
> I have not made PRs to Julia for some time. Could someone please guide me through which CI jobs are running the actual tests of the functionality? (as there seem to be many jobs doing some specific testing) Thank you!

AFAIK "package" jobs only build Julia, "tester" jobs run the tests. Not sure about others.

JeffBezanson commented:
Looks great, just need to fix the tests.

bkamins committed:
improve implementation in corner cases

bkamins committed:
code fixes

bkamins commented on base/bitarray.jl:
```diff
@@ -987,25 +988,68 @@ function deleteat!(B::BitVector, inds)
 
     (p, s) = y
     checkbounds(B, p)
+    p isa Bool && throw(ArgumentError("invalid index $p of type Bool"))
```
We can throw an error here (and not just warn about deprecation), because earlier it already threw an error as a side effect of lack of conversion of `p` to `Int` where needed later. (same below)

The current behavior is:
```
julia> deleteat!(falses(5), Any[true])
ERROR: MethodError: no method matching copy_chunks!(::Vector{UInt64}, ::Bool, ::Vector{UInt64}, ::Int64, ::Int64)

julia> deleteat!(falses(5), Any[1, true])
ERROR: ArgumentError: indices must be unique and sorted
```
Now in both cases we will throw `throw(ArgumentError("invalid index $p of type Bool"))`

bkamins commented on base/bitarray.jl:
```diff
@@ -1020,6 +1064,10 @@ function deleteat!(B::BitVector, inds)
 end
 
 function splice!(B::BitVector, i::Integer)
+    # TODO: after deprecation remove the four lines below
```
I have also added the checks for `Bool` to `splice!` as they are done in non-`BitVector` case:
```
julia> splice!([1,2,3], true)
ERROR: ArgumentError: invalid index: true of type Bool

julia> splice!([1,2,3], true, [4,5])
ERROR: ArgumentError: invalid index: true of type Bool
```

bkamins commented:
> just need to fix the tests.

I have fixed. Actually the test was correct, but it was failing because there was an additional issue with lacking conversion to `Int` in the old code that required fixing. Now it should be good for a review. I have added some comments in places that are non-obvious without running the code.

bkamins commented:
re-triggering CI

bkamins commented:
A single CI error seems unrelated, so this should be good for a final review.

JeffBezanson commented on base/array.jl:
```diff
@@ -1462,12 +1462,22 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 2)
 """
-deleteat!(a::Vector, i::Integer) = (_deleteat!(a, i, 1); a)
+function deleteat!(a::Vector, i::Integer)
+    i isa Bool && depwarn("passing Bool as an index is deprecated", deleteat!)
```
```suggestion
    i isa Bool && depwarn("passing Bool as an index is deprecated", :deleteat!)
```

JeffBezanson commented on base/array.jl:
```diff
@@ -1462,12 +1462,22 @@ julia> deleteat!([6, 5, 4, 3, 2, 1], 2)
+    else
+        n = length(a)
+        f = first(r)
+        f isa Bool && depwarn("passing Bool as an index is deprecated", deleteat!)
```
```suggestion
        f isa Bool && depwarn("passing Bool as an index is deprecated", :deleteat!)
```

JeffBezanson commented on base/bitarray.jl:
```diff
@@ -947,6 +947,7 @@ function _deleteat!(B::BitVector, i::Int)
 end
 
 function deleteat!(B::BitVector, i::Integer)
+    i isa Bool && depwarn("passing Bool as an index is deprecated", deleteat!)
```
```suggestion
    i isa Bool && depwarn("passing Bool as an index is deprecated", :deleteat!)
```

JeffBezanson commented on base/bitarray.jl:
```diff
@@ -1020,6 +1064,10 @@ function deleteat!(B::BitVector, inds)
+    # TODO: after deprecation remove the four lines below
+    #       as v = B[i] is enough to do both bounds checking
+    #       and Bool check then just pass Int(i) to _deleteat!
+    i isa Bool && depwarn("passing Bool as an index is deprecated", splice!)
```
```suggestion
    i isa Bool && depwarn("passing Bool as an index is deprecated", :splice!)
```

JeffBezanson commented on base/bitarray.jl:
```diff
@@ -1032,8 +1080,10 @@ end
 const _default_bit_splice = BitVector()
 
 function splice!(B::BitVector, r::Union{AbstractUnitRange{Int}, Integer}, ins::AbstractArray = _default_bit_splice)
+    r isa Bool && depwarn("passing Bool as an index is deprecated", splice!)
```
```suggestion
    r isa Bool && depwarn("passing Bool as an index is deprecated", :splice!)
```

JeffBezanson committed:
Apply suggestions from code review

bkamins commented:
Could this PR be backported to 1.6 release? see https://github.com/JuliaData/DataFrames.jl/issues/3300
